### PR TITLE
Enxr ssh

### PR DESCRIPTION
**Pull request by ashojosh**
Added EnXR support to PyATS

### CODE REVIEW

File: src/unicon/plugins/iosxr/enxr/connection_provider.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+__author__ = "Ashok Joshi <ashojosh@cisco.com>"
+
+import traceback
+import pexpect
```
Comment (LenzDong) [inline src/unicon/plugins/iosxr/enxr/connection_provider.py:4]: Could you please reuse the Spawn class in unicon
https://pubhub.devnetcloud.com/media/unicon/docs/developer_guide/eal.html#spawn
```
from unicon.eal.expect import Spawn

File: src/unicon/plugins/iosxr/enxr/connection_provider.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            if not self.connected:
+                p = pexpect.spawn(self.spawn)
+                p.expect(self.prompt, timeout=10)
+                self.connected = True
+                self.ssh = p
+            return 'Connected with %s' % (self.connection.device.name)
+        except pexpect.exceptions.TIMEOUT:
+            log.error('EnXR connect failed ')
+        except Exception:
+            log.error('EnXR connect failed ')
```
Comment (LenzDong) [inline src/unicon/plugins/iosxr/enxr/connection_provider.py:56]: please use `self.connection.log`

File: src/unicon/plugins/iosxr/enxr/service_implementation.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    def call_service(self, cmd, dialog=Dialog([]),
+                     timeout=20, *args, **kwargs):
+        self.connect()
+        return self._send(cmd)
+
+    def post_service(self, *args, **kwargs):
+        pass
+
+    @property
+    def connected(self):
```
Comment (LenzDong) [inline src/unicon/plugins/iosxr/enxr/service_implementation.py:40]: this should be a `Connection` object attribute?

File: src/unicon/plugins/iosxr/enxr/service_implementation.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        """ Return True if session is connected."""
+        return self._connected
+
+    @connected.setter
+    def connected(self, is_connected):
+        self._connected = is_connected
+
+    def connect(self, **kwargs):
```
Comment (LenzDong) [inline src/unicon/plugins/iosxr/enxr/service_implementation.py:48]: `connect()` function should be in connection provider, not inside service

File: src/unicon/plugins/iosxr/enxr/connection_provider.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import traceback
+import pexpect
+from unicon.plugins.iosxr.connection_provider \
+    import IOSXRSingleRpConnectionProvider
+from unicon.plugins.iosxr.enxr.settings import EnxrSettings
+from unicon import log
+
+
+class IOSXREnxrSingleRpConnectionProvider(IOSXRSingleRpConnectionProvider):
```
Comment (LenzDong) [inline src/unicon/plugins/iosxr/enxr/connection_provider.py:11]: please refer to this guide line to implement new plugins:
https://pubhub.devnetcloud.com/media/unicon/docs/developer_guide/plugins.html#contribution

**Comment by LenzDong** [comment]
Please add unit test for the new plugin

File: src/unicon/plugins/iosxr/enxr/service_implementation.py (hunk 4 above)
Comment (ashojosh) [reply src/unicon/plugins/iosxr/enxr/service_implementation.py:48]: Done. 

I have basically referred connection_provider object here to get & use the spawn.

File: src/unicon/plugins/iosxr/enxr/service_implementation.py (hunk 3 above)
Comment (ashojosh) [reply src/unicon/plugins/iosxr/enxr/service_implementation.py:40]: Sorry, I did not get it. I have implemented the BaseService class. Is there anything that  I am missing?

File: src/unicon/plugins/iosxr/enxr/connection_provider.py (hunk 2 above)
Comment (ashojosh) [reply src/unicon/plugins/iosxr/enxr/connection_provider.py:56]: Done

File: src/unicon/plugins/iosxr/enxr/connection_provider.py (hunk 1 above)
Comment (ashojosh) [reply src/unicon/plugins/iosxr/enxr/connection_provider.py:4]: using Spawn class is creating an issue in service_implementation. That's why I have used pexpect.spawn.

In the service implementation file, 'con.ssh.before' is giving an error that the Spawn class has no attribute before.

**Comment by ashojosh** [comment]
Sorry for the delayed response to the pull request comments. My CEC ID had been locked because of some reason.

**Comment by miott** [comment]
@LenzDong - Ashok has addressed review comments.  Can this pull request get merged or is there more he must do?

File: src/unicon/plugins/iosxr/enxr/connection_provider.py (hunk 1 above)
Comment (simingy) [reply src/unicon/plugins/iosxr/enxr/connection_provider.py:4]: then we should fix that issue. we are not going to start introducing another dependency (`pexpect`).

Please work with Lunzhi to have it resolved

**Review by simingy — CHANGES_REQUESTED** [review]

File: src/unicon/plugins/iosxr/enxr/connection_provider.py (hunk 1 above)
Comment (LenzDong) [reply src/unicon/plugins/iosxr/enxr/connection_provider.py:4]: Could you share what's the issue when using the Spawn class in unicon?

**Comment by LenzDong** [comment]
As discussed, you can use the default iosxr plugin no need to create a new one for enxr

**Comment by miott** [comment]
> As discussed, you can use the default iosxr plugin no need to create a new one for enxr

Can you give an example of the device settings in the testbed file?  Not sure what you mean by "default".

**Comment by LenzDong** [comment]
> > As discussed, you can use the default iosxr plugin no need to create a new one for enxr
> 
> Can you give an example of the device settings in the testbed file? Not sure what you mean by "default".

Here is an example testbed
```
devices:
    # WARNING: device name must match configured hostname
    bgl-ads-4861:
        type: iosxr
        os: iosxr   # <<< use the default iosxr plugin
        alias: uut
        credentials:
            default:
                username: admin
                password: admin
        connections:
            cli:
                class: unicon.Unicon
                command: exec   # <<< provide this to specify the spawn command instead of protocol and ip
```

**Comment by miott** [comment]
> > > As discussed, you can use the default iosxr plugin no need to create a new one for enxr
> > 
> > 
> > Can you give an example of the device settings in the testbed file? Not sure what you mean by "default".
> 
> Here is an example testbed
> 
> ```
> devices:
>     # WARNING: device name must match configured hostname
>     bgl-ads-4861:
>         type: iosxr
>         os: iosxr   # <<< use the default iosxr plugin
>         alias: uut
>         credentials:
>             default:
>                 username: admin
>                 password: admin
>         connections:
>             cli:
>                 class: unicon.Unicon
>                 command: exec   # <<< provide this to specify the spawn command instead of protocol and ip
> ```

How do you use it? It does not have standard APIs like "configure" and "execute"
```
>>> dir(dev)
['CFG_TOPOLOGY_CLASS', 'ENTRY_POINT_NAME', '__class__', '__class__', '__contains__', '__delattr__', '__delattr__', '__device__', '__dict__', '__dict__', '__dir__', '__dir__', '__doc__', '__doc__', '__eq__', '__eq__', '__format__', '__format__', '__ge__', '__ge__', '__getattr__', '__getattr__', '__getattribute__', '__getattribute__', '__gt__', '__gt__', '__hash__', '__hash__', '__init__', '__init__', '__init_subclass__', '__init_subclass__', '__iter__', '__le__', '__le__', '__lt__', '__lt__', '__module__', '__module__', '__ne__', '__ne__', '__new__', '__new__', '__reduce__', '__reduce__', '__reduce_ex__', '__reduce_ex__', '__repr__', '__repr__', '__setattr__', '__setattr__', '__sizeof__', '__sizeof__', '__str__', '__str__', '__subclasshook__', '__subclasshook__', '__testbed__', '__weakref__', '__weakref__', '_default_connection_cls', '_deprecation_check', 'add_interface', 'alias', 'clean', 'connect', 'connectionmgr', 'connections', 'connections', 'credentials', 'custom', 'default_alias', 'default_connection_cls', 'destroy', 'destroy_all', 'device', 'disconnect', 'disconnect_all', 'find_links', 'instantiate', 'instantiate_pool', 'interfaces', 'is_connected', 'links', 'name', 'os', 'passwords', 'remote_devices', 'remote_interfaces', 'remove_interface', 'start_pool', 'tacacs', 'testbed', 'type']
>>> dev.type
'iosxr'
>>> dev.os
'iosxr'
>>>
```

**Comment by LenzDong** [comment]
You will need to connect the device first
```
dev.connect(via='cli')

**Pull request closed without merging**